Handle direct CanncelationException on timeout in JdkClientHttpRequest

giampa91 opened the pull request:
CancellationExceptions are thrown instead of the expected ResourceAccessException during timeout scenarios. 
Handle the CancellationExceptions in order to throw an ResourceAccessException when timeout occurred

Closes gh-33973

giampa91 committed:
CancellationExceptions are thrown instead of the expected ResourceAccessException during timeout scenarios.
Handle the CancellationExceptions in order to throw an ResourceAccessException when timeout occurred

Closes gh-33973

Signed-off-by: giampaolo <giampaorr@gmail.com>

fix: use timeoutHandler with a flag isTimeout

    Closes gh-33973

    Signed-off-by: giampaolo <giampaorr@gmail.com>

rstoyanchev reviewed (changes requested):
There seems to be no purpose to moving the TimeoutHandler

giampa91 commented on spring-web/src/main/java/org/springframework/http/client/JdkClientHttpRequest.java:
```diff
@@ -224,6 +228,7 @@ public ByteBuffer map(byte[] b, int off, int len) {
 	private static final class TimeoutHandler {
 
 		private final CompletableFuture<Void> timeoutFuture;
+		private boolean isTimeout=false;
```
the flag is used in line 240 and is setted to true, and on default is false

giampa91 commented on spring-web/src/main/java/org/springframework/http/client/JdkClientHttpRequest.java:
```diff
@@ -96,12 +96,13 @@ public URI getURI() {
 	@Override
 	protected ClientHttpResponse executeInternal(HttpHeaders headers, @Nullable Body body) throws IOException {
 		CompletableFuture<HttpResponse<InputStream>> responseFuture = null;
+		TimeoutHandler timeoutHandler = null;
```
this change was to make the TimeoutHandler readable outside the try

giampa91 commented:
Thank you for your patience! After a careful review, I realized I hadn’t double-checked the code thoroughly enough.

I’ve now updated the implementation to better distinguish between a cancelled request and a genuine timeout by explicitly checking the timeout variable.

Additionally, I’ve added tests covering both scenarios. While there might still be room for optimization, I believe this approach reflects real-world usage and edge cases.

rstoyanchev reviewed (approved):
Okay the intent is more clear now.
